File: app/src/main/java/network/minter/bipwallet/exchange/adapters/CoinsListAdapter.java
/*
 * Copyright (C) by MinterTeam. 2018
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

package network.minter.bipwallet.exchange.adapters;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import network.minter.bipwallet.R;
import network.minter.explorer.models.CoinItem;
import timber.log.Timber;

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich <edward.vstock@gmail.com>
 * @TODO remove adapter duplication, create base autocompletion adapter
 */
public class CoinsListAdapter extends ArrayAdapter<CoinItem> implements Filterable {
    private List<CoinItem> mItems;
    private List<CoinItem> mItemsAll;
    private List<CoinItem> mSuggestions;
    private int mViewResourceId;
    private LayoutInflater mInflater;
    private OnItemClickListener mOnItemClickListener;

    public CoinsListAdapter(@NonNull Context context, @NonNull List<CoinItem> items) {
        super(context, R.layout.search_item, items);
        setItems(items);
        mViewResourceId = R.layout.search_item;
    }

    public void setOnItemClickListener(OnItemClickListener listener) {
        mOnItemClickListener = listener;
    }

    @NonNull
    public View getView(int position, View convertView, @NonNull ViewGroup parent) {
        View v = convertView;
        ViewHolder holder;
        if (mInflater == null) {
            mInflater = LayoutInflater.from(parent.getContext());
        }

        if (v == null) {
            v = mInflater.inflate(mViewResourceId, null);
            holder = new ViewHolder(v);
            v.setTag(holder);
        } else {
            holder = ((ViewHolder) v.getTag());
        }

        onBindViewHolder(holder, position);
        return v;
    }

    @NonNull
    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            public String convertResultToString(Object resultValue) {
                return ((CoinItem) (resultValue)).symbol;
            }

            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                if (constraint != null) {
                    mSuggestions.clear();
                    for (CoinItem item : mItemsAll) {
                        if (item.symbol != null && item.symbol.toLowerCase().startsWith(constraint.toString().toLowerCase())) {
                            mSuggestions.add(item);
                        } else if (item.name != null && item.name.toLowerCase().startsWith(constraint.toString().toLowerCase())) {
                            mSuggestions.add(item);
                        }
                    }
                    FilterResults filterResults = new FilterResults();
                    filterResults.values = mSuggestions;
                    filterResults.count = mSuggestions.size();
                    return filterResults;
                } else {
                    return new FilterResults();
                }
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                //noinspection unchecked
                List<CoinItem> filteredList = (List<CoinItem>) results.values;
                //noinspection ConstantConditions
                if (results != null && results.count > 0) {
                    Timber.d("Add filter item (items: %d)", results.count);
                    clear();
                    for (CoinItem c : filteredList) {
                        new Handler(Looper.getMainLooper()).post(() -> add(c));
                    }
                    notifyDataSetChanged();
                }
            }
        };
    }

    public void setItems(List<CoinItem> items) {
        mItems = items;
        //noinspection unchecked
        mItemsAll = (ArrayList<CoinItem>) ((ArrayList<CoinItem>) mItems).clone();
        mSuggestions = new ArrayList<>();
    }

    private void onBindViewHolder(ViewHolder vh, int position) {
        if (mOnItemClickListener != null) {
            vh.itemView.setOnClickListener(vv -> mOnItemClickListener.onClick(mItems.get(position), position));
        }

        CoinItem item = mItems.get(position);

        vh.title.setText(item.symbol);
        vh.subtitle.setVisibility(View.VISIBLE);
        vh.subtitle.setText(item.name);
    }

    public interface OnItemClickListener {
        void onClick(CoinItem item, int position);
    }

    static class ViewHolder {
        View itemView;
        @BindView(R.id.search_item_title) TextView title;
        @BindView(R.id.search_item_subtitle) TextView subtitle;

        public ViewHolder(View itemView) {
            this.itemView = itemView;
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/network/minter/bipwallet/exchange/adapters/CoinsListAdapter.kt
/*
 * Copyright (C) by MinterTeam. 2020
 * @link <a href="https://github.com/MinterTeam">Org Github</a>
 * @link <a href="https://github.com/edwardstock">Maintainer Github</a>
 *
 * The MIT License
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package network.minter.bipwallet.exchange.adapters

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.Filter
import android.widget.Filterable
import network.minter.bipwallet.R
import network.minter.bipwallet.databinding.SearchItemBinding
import network.minter.bipwallet.internal.helpers.ViewExtensions
import network.minter.bipwallet.internal.helpers.ViewExtensions.visible
import network.minter.explorer.models.CoinItem
import java.util.*

/**
 * minter-android-wallet. 2018
 * @author Eduard Maximovich <edward.vstock></edward.vstock>@gmail.com>
 * @TODO remove adapter duplication, create base autocompletion adapter
 */
class CoinsListAdapter(context: Context, items: List<CoinItem>) : ArrayAdapter<CoinItem>(context, R.layout.search_item, items), Filterable {
    private var mItemsAll: List<CoinItem>? = null
    private var mSuggestions: MutableList<CoinItem>? = null
    private val mViewResourceId: Int
    private var mInflater: LayoutInflater? = null
    private var mOnItemClickListener: ((CoinItem, Int) -> Unit)? = null

    init {
        setItems(items)
        mViewResourceId = R.layout.search_item
    }

    fun setOnItemClickListener(listener: (CoinItem, Int) -> Unit) {
        mOnItemClickListener = listener
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var v = convertView
        val holder: ViewHolder
        if (mInflater == null) {
            mInflater = LayoutInflater.from(parent.context)
        }
        if (v == null) {
            v = mInflater!!.inflate(mViewResourceId, null)
            holder = ViewHolder(SearchItemBinding.bind(v))
            v.tag = holder
        } else {
            holder = v.tag as ViewHolder
        }
        onBindViewHolder(holder, position)
        return v!!
    }

    private val publishLock = Any()
    override fun getFilter(): Filter {
        return object : Filter() {
            override fun convertResultToString(resultValue: Any): String {
                return (resultValue as CoinItem).symbol
            }

            override fun performFiltering(constraint: CharSequence?): FilterResults {
                return if (constraint != null) {
                    mSuggestions!!.clear()
                    for (item in mItemsAll!!) {
                        if (item.symbol != null && item.symbol.toLowerCase().startsWith(constraint.toString().toLowerCase())) {
                            mSuggestions!!.add(item)
                        }
                    }
                    val filterResults = FilterResults()
                    filterResults.values = mSuggestions
                    filterResults.count = mSuggestions!!.size
                    filterResults
                } else {
                    FilterResults()
                }
            }

            @Suppress("UNCHECKED_CAST")
            override fun publishResults(constraint: CharSequence?, results: FilterResults?) {
                val filteredList = results?.values as List<CoinItem>?
                synchronized(publishLock) {
                    if (results != null && filteredList != null && results.count > 0) {
                        clear()
                        for (c in filteredList) {
                            add(c)
                        }
                        notifyDataSetChanged()
                    } else {
                        clear()
                        notifyDataSetInvalidated()
                    }
                }
            }
        }
    }

    fun setItems(items: List<CoinItem>) {
        mItemsAll = items
        mSuggestions = ArrayList()
    }

    private fun onBindViewHolder(vh: ViewHolder, position: Int) {
        vh.b.root.setOnClickListener { mOnItemClickListener?.invoke(getItem(position)!!, position) }

        ViewExtensions.listItemBackgroundRippleRounded(vh.b.root, position, count)

        vh.b.separator.visible = !(count == 1 || position == count - 1)

        val item = getItem(position)!!
        vh.b.searchItemTitle.text = item.symbol
        vh.b.searchItemSubtitle.visible = true
        vh.b.searchItemSubtitle.text = item.name
    }

    interface OnItemClickListener {
        fun onClick(item: CoinItem, position: Int)
    }

    class ViewHolder(val b: SearchItemBinding)
}